Improving `backup_and_restore.sh` script before new-coming features

h3ssan opened the pull request:
<!-- _Please make sure to review and check all of these items, otherwise we might refuse your PR:_ -->

## Contribution Guidelines

* [x] I've read the [contribution guidelines](https://github.com/mailcow/mailcow-dockerized/blob/master/CONTRIBUTING.md) and wholeheartedly agree them

## What does this PR include?

- [Adding `print_usage` function to `backup_and_restore.sh` script](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/e3311aef9470230c876aaeb04020093f1401c058)
- [Add colored messages in `backup_and_restore.sh` script](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/2460a98f20b0fa5a13a67489234f4045e2ab0017)
- [Move the `thread count notice` into the `print_usage`](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/5eceecf8cc50f8ea60a9bd6e98a954fd96bb6d96)
- [Fix: Double quoted variable `BACKUP_LOCATION`.](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/82def75ec557fdb2d7f4066100466c1f2f8fffda) - **Critical**: Since it might has spaces therefore will corrupt the backup/restore process.
- [Bug Fix: overwriting filesystem possible](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/a621aea4510ae39703133f09a1d7f9b0aea4db2e) - **Critical**: Previous code was not checking if the path is a `symbolic` to a filesystem, if user is root and the given path is a `symbolic` of filesystem then it will overwrite it!
- [Simplify the check expression of THREADS](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/0d95962f1c09b07bd07eacdad167b7723901ce72)
- [Implement check_required_tools function](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/67ba81a41c7c1d6181ed373f0862f5eca4593af2) - Implement `check_required_tools` function in `backup_and_restore.sh` script, as well call it in the beginning of the script instead of at the end
- [Clarifying the duplicate assigning of MAILCOW_BACKUP_LOCATION](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/f64efb3b1dd5f46de22775f8884d8bd8229d5679)
- [Fix Bug: backup_and_restore.sh script, increment i only if needed](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/cfe13b0d9e5f9691e4ba3d74bcb44bb06549fbf5)
- [Fix: Crash on entering non-numeric value](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/4d00ca6dbfa92934013de4fe448055f354292dcf)
- [Fix: backup_location should be exists, for the restore](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/a2ab5271bc4e5a1c9f5fb31996024c006210d037)
- [Fix: Crash on entering non-numeric value](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/ea24505f35d4b1022865a77e06ff57c1aa0011b2)
- [Implement declare_restore_point function](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/c45b860f561bf8d8da8eb9bbb2250a956760773f)
- [Typo in calling declare_restore_point](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/de743cd2bf11461fa8b9474ba72b1f1a81302ba1)
- [Fix: grep: warning: stray \ before :](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/728a9951aff5f942ee691311adaaaa0152f4c2da)
- [Implement declare_file_selection function](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/d398ecaf0890a0bb28355e3fc4d925d9ee30c2b2)
- [Implement restore_docker_component to restore components in one place](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/bd229b9dbc5fb7b8d3f8dd4974966d49e734e978)
- [Fix last grep: warning: stray \ before :](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/d0b38b7910c3962219aa8789aca871fe6dbfb0cb)
- [Write a clarification comment for function restore_docker_component](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/93fc601b0b607c54b6724d5dc749bc1decd31217)
- [Introduce new help screen, with new flags depends script](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/78a62663ce90c63a67ccf602c64670c0e9cd0481)

### Short Description

This PR is not finished yet, I'll include many re-write code for the `backup_and_restore.sh` script. Enhancement, refactor and improved code, before starting to implement a new features.

###  Affected Containers

None

## Did you run tests?

Tested manually

### What did you tested?

- `backup_and_restore.sh` script

### What were the final results? (Awaited, got)

All good

h3ssan committed:
Adding `print_usage` function to `backup_and_restore.sh` script

h3ssan committed:
Add colored messages in `backup_and_restore.sh` script

h3ssan committed:
Move the `thread count notice` into the usage menu

h3ssan committed:
Fix: Double quoted variable `BACKUP_LOCATION`.

Since it might have spaces, it must be double quoted.

h3ssan committed:
Bug Fix: overwriting filesystem possible

Previous code was not checking if the path is a `symbolic` to a filesystem, if user is root and the given path is a `symbolic` of filesystem then it will overwrite it!

h3ssan committed:
Simplify the check expression of `THREADS`

h3ssan committed:
Implement `check_required_tools` function

Implement `check_required_tools` function in `backup_and_restore.sh` script, as well call it in the beginning of the script instead of at the end

h3ssan commented:
Hey @DerLinkman, you can now resolve this PR, since I will need a few time to refactor and fully test the `backup` and `restore` functionality, since then you can resolve this PR now to avoid conflicts.

DerLinkman commented:
> Hey @DerLinkman, you can now resolve this PR, since I will need a few time to refactor and fully test the `backup` and `restore` functionality, since then you can resolve this PR now to avoid conflicts.

Hello @h3ssan! Is this pr now ready to merge or not? If not, we can convert it to a draft.

h3ssan commented:
Hey there @DerLinkman , yes it’s ready to be merged.

still there’s 2 functions I need to look into but might take some time so yes, resolve it

DerLinkman reviewed (approved):
Looks fine!

h3ssan committed:
Clarifying the duplicate assigning of `MAILCOW_BACKUP_LOCATION`

h3ssan committed:
Fix Bug: `backup_and_restore.sh` script, increment `i` only if needed

h3ssan commented:
## Clarification

Commit [`Fix Bug: backup_and_restore.sh script, increment i only if needed`](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/cfe13b0d9e5f9691e4ba3d74bcb44bb06549fbf5), why?

### Original Code

```bash
  i=1 # Starts at 1, even before finding any folder
  declare -A FOLDER_SELECTION
  if [[ $(find "${BACKUP_LOCATION}"/mailcow-* -maxdepth 1 -type d 2> /dev/null| wc -l) -lt 1 ]]; then
    echo -e "\e[31mSelected backup location has no subfolders\e[0m"
    exit 1
  fi
  for folder in $(ls -d "${BACKUP_LOCATION}"/mailcow-*/); do
    echo "[ ${i} ] - ${folder}"
    FOLDER_SELECTION[${i}]="${folder}"
    ((i++)) 
  done

```

If `i` is initial to `1`,  this will cause:

- `i = 1` - Initial
- Checking if there's a folder, then increment `i`, therefore `1 = 2`

Here we have `i` equals to `2` but we only found one folder !
Then, when the user choose which folder to restore, the user can simply choose `2` not `1` and the script will continue working, check the following image:

![](https://i.imgur.com/rK4XDor.png)

h3ssan committed:
Fix: Crash on entering non-numeric value

`backup_and_restore.sh` script, with `restore` flag, crashes when entering a non-numeric value as backup choice

h3ssan commented:
## Clarification 

Commit: [Fix: Crash on entering non-numeric value](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/4d00ca6dbfa92934013de4fe448055f354292dcf), fixes the following issue:

Enter `,` (non-numeric value, actually) will case a crash:

![](https://i.imgur.com/awTLYym.png)

h3ssan committed:
Fix: backup_location should be exists, for the restore

h3ssan committed:
Fix: Crash on entering non-numeric value

Fix the crash on entering non-numeric value, as well as make the code more readable

h3ssan commented:
Till this point, what I have tested is below:

| Tested Command | Expected | Status |
| --- | --- | --- |
| `./backup_and_restore.sh backup vmail` | Backed up<sup>`1`</sup> | Good &#10003;  |
| `./backup_and_restore.sh backup all` | Backed up<sup>`1`</sup> | Good &#10003;  |
| `./backup_and_restore.sh backup anythingelse` | Throw `Unknown argument: anythingelse` | Good &#10003;  |
| `./backup_and_restore.sh anythingelse` | print the usage | Good &#10003;  |
| `./backup_and_restore.sh restore` with correct folder | Restored<sup>`1`</sup> | Good &#10003;  |
| `./backup_and_restore.sh restore` with non-existence folder | Throw `location not found` | Good &#10003;  |
| `./backup_and_restore.sh restore` with location with no data | Throw `no subfolders` | Good &#10003;  |
| `./backup_and_restore.sh restore` with location to `/dev/sda` | Throw `invalid path` | Good &#10003;  |
| `./backup_and_restore.sh backup vmail` with location to `/dev/sda` | Throw `invalid path` | Good &#10003;  |

---

<sup>`1`</sup> I DID NOT check the actual process, just trusted what the script said! Therefore it might need extensive/dedicated testing.

h3ssan committed:
Implement `declare_restore_point` function

Refactor `backup_and_restore.sh` script, implement `declare_restore_point` function.

h3ssan commented:
Sorry about the last commit, there was an issue with the `committer` and `author` details. I soft reset, fixed it and push again.

h3ssan commented:
Please note that this commit [`Implement declare_restore_point function`](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/c45b860f561bf8d8da8eb9bbb2250a956760773f) still need testing. **I'll do the test in a few hours.**

h3ssan committed:
Typo in `calling declare_restore_point`

h3ssan committed:
Fix: `grep: warning: stray \ before :`

h3ssan committed:
Implement `declare_file_selection` function

Refactor `backup_and_restore.sh` script, implement `declare_file_selection` function

h3ssan committed:
Implement `restore_docker_component` to restore components in one place

Implement `restore_docker_component` function to restore `vmail`, `redis`, `crypt`, `rspamd` and `postfix`, all from one function

h3ssan committed:
Fix last `grep: warning: stray \ before :`

h3ssan committed:
Write a clarification comment for function `restore_docker_component`

h3ssan committed:
Introduce new `help` screen, with new flags depends script

- Add New help screen/menu
- Script now works with flags, such as `-r|--restore`, `-b|--backup` and so on.
- Add new feature for restoring many components at once.
and many more.

h3ssan commented:
What to be expected in the commit: [`Introduce new help screen, with new flags depends script`](https://github.com/mailcow/mailcow-dockerized/pull/6030/commits/78a62663ce90c63a67ccf602c64670c0e9cd0481)

## Features

- Restore multiple components at once, such as `./backup_and_restore.sh --restore /path/to/restore -c vmail -c crypt -c mysql` or even using `./backup_and_restore.sh --restore /path/to/restore -c all` to restore everything at once.
- Making the script as flexible as possible.

## Enhancement

- Introduce a new help menu, check it out with `./backup_and_restore.sh --help`
- New flag system, also check it with `./backup_and_restore.sh --help`
- Define `MAILCOW_BACKUP_LOCATION` in your `~/.bashrc` to shorten time ! then just do `./backup_and_restore.sh -c all` to backup everything without specifying --backup flag!
- Use named colors, instead of typing a coded color value.
- Add many comments in the script to avoid overwhelmed code.
-  Avoid nested statements.
- Big improvement in `--delete-days` flag.

h3ssan committed:
Use named colors, instead of ANSI codes

h3ssan committed:
Fix: THREADS not printed in `backup_and_restore.sh`

h3ssan committed:
Fix: Double quote array expansions to avoid re-splitting elements

DerLinkman reviewed (changes requested):
Thank you! Looks good so far.

A few things i noticed:

1. If you run a restore job (e.g: ./helper-scripts/backup_and_restore.sh -t 4 -r /opt/backup -c mysql) then the section:

```
Using 4 thread(s) for this run.
Found project name mailcowdockerized
Using /opt/backup as restore location...
[ 1 ] - /opt/backup/mailcow-2024-08-24-08-35-01/
[ 2 ] - /opt/backup/mailcow-2024-08-25-08-35-01/
[ 3 ] - /opt/backup/mailcow-2024-08-26-08-35-01/
[ 4 ] - /opt/backup/mailcow-2024-08-27-08-14-51/

Select a restore point: 4


Matching available components to restore:  
here is nothing displayed
```

Where on the otherhand if you select other types like vmail, rspamd etc. those are printed. It seems mysql is missing when you manually select this here.

If you select all as component type then mysql/mariadb is shown in the list

Btw: What are the reason of these numbers for? As you now cannot select them anymore? Is that the order they get restored now?

This i mean:

```
Matching available components to restore:
[ 1 ] - crypt
[ 2 ] - rspamd
[ 3 ] - vmail
[ 4 ] - redis
[ 5 ] - postfix
[ 6 ] - mariadb
```

DerLinkman reviewed (commented):
> Thank you! Looks good so far.

> A few things i noticed:

> 1. If you run a restore job (e.g: ./helper-scripts/backup_and_restore.sh -t 4 -r /opt/backup -c mysql) then the section:

> ```
Using 4 thread(s) for this run.
Found project name mailcowdockerized
Using /opt/backup as restore location...
[ 1 ] - /opt/backup/mailcow-2024-08-24-08-35-01/
[ 2 ] - /opt/backup/mailcow-2024-08-25-08-35-01/
[ 3 ] - /opt/backup/mailcow-2024-08-26-08-35-01/
[ 4 ] - /opt/backup/mailcow-2024-08-27-08-14-51/

> Select a restore point: 4

> 
Matching available components to restore:  
here is nothing displayed
```

> Where on the otherhand if you select other types like vmail, rspamd etc. those are printed. It seems mysql is missing when you manually select this here. 

If you select all as components, mariadb/mysql is shown in the list

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,621 @@
-  echo "Backup directory needs to be given as absolute path (starting with /)."
-  exit 1
-fi
+  if grep --help 2>&1 | head -n 1 | grep -q -i "busybox"; then
```
I would add grep to the list on required_tools (line 30 too)

h3ssan replied:
Done.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,621 @@
+  local BACKUP_LOCATION="${1}"
+
+  # Check subfolders inside BACKUP_LOCATION
+  if [[ $(find "${BACKUP_LOCATION}"/mailcow-* -maxdepth 1 -type d 2> /dev/null| wc -l) -lt 1 ]]; then
```
I don't know for sure, if find has a similar behavior like grep on busybox or if it is the same for both?

h3ssan replied:
Yes, searched for it. `find` has BusyBox version too. Added to the check list for BusyBox.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,621 @@
-        -v ${BACKUP_LOCATION}/mailcow-${DATE}:/backup:z \
+        -v "${MAILCOW_BACKUP_LOCATION}"/mailcow-${DATE}:/backup:z \
         -v $(docker volume ls -qf name=^${CMPS_PRJ}_vmail-vol-1$):/vmail:ro,z \
-        ${DEBIAN_DOCKER_IMAGE} /bin/tar --warning='no-file-ignored' --use-compress-program="pigz --rsyncable -p ${THREADS}" -Pcvpf /backup/backup_vmail.tar.gz /vmail
```
tar is also needed to check in the required tools (line 30) then

pigz too i assume?

h3ssan replied:
Both `tar` and `pigz` are running from inside `mailcow/backup:latest` which is doing only one thing:

```Dockerfile
FROM debian:bookworm-slim

RUN apt update && apt install pigz
```

so no need to be checked in the host machine, right?

DerLinkman replied:
Ah yes you're right... myself created this though :D No no need for action here.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,621 @@
+          fi
+          echo "Modifying mailcow.conf..."
+          source ${RESTORE_LOCATION}/mailcow.conf
+          sed -i --follow-symlinks "/DBNAME/c\DBNAME=${DBNAME}" ${SCRIPT_DIR}/../mailcow.conf
```
sed too (line 30) also check if sed is not busybox, as they have differences in command syntax.

h3ssan replied:
Done.

DerLinkman reviewed (changes requested):
See comments

h3ssan commented:
> ```
> Using 4 thread(s) for this run.
> Found project name mailcowdockerized
> Using /opt/backup as restore location...
> [ 1 ] - /opt/backup/mailcow-2024-08-24-08-35-01/
> [ 2 ] - /opt/backup/mailcow-2024-08-25-08-35-01/
> [ 3 ] - /opt/backup/mailcow-2024-08-26-08-35-01/
> [ 4 ] - /opt/backup/mailcow-2024-08-27-08-14-51/
> 
> Select a restore point: 4
> 
> 
> Matching available components to restore:  
> here is nothing displayed
> ```
> 
> Where on the otherhand if you select other types like vmail, rspamd etc. those are printed. It seems mysql is missing when you manually select this here.
> 
> If you select all as component type then mysql/mariadb is shown in the list
>

Okay, so the `mysql` is not shown, since there's mix in use between `mysql` and `mariadb` but the actual database is `mariadb` and the actual backup file is `backup_mariadb.tar.gz` therefore **I will completely change `mysql` to `mariadb`.**
Are you ok with this change?

**Edit: I do not know if it was previously storing backups as `mysql` instead of now `mariadb`, so, I will just fix the issue instead of changing everything to `mariadb`.**

h3ssan commented:
> Btw: What are the reason of these numbers for? As you now cannot select them anymore? Is that the order they get restored now?

The reason of the numbers is for:
1. Ordering - which going to be restored in that sequence.
2. Identifying - so you'll know which components are going to be restored, because you might choose to restore `vmail` on a restore path that did not backup vmail at all.

Example:
If you backup only the `crypt`, and then you choose to restore the `vmail`, it will now show up, since there's no `vmail` in the restore directory you specified.

h3ssan committed:
Fix: mysql component not shown when trying to restore it

h3ssan committed:
Fix: Make sure there's components before restoring process

h3ssan committed:
Add `grep`, `find` and `sed` to the required_tools, also BusyBox check.

h3ssan committed:
Remove duplicate threads notice for backup process

h3ssan commented:
Thank you @DerLinkman for the testing process.
I have fixed the issues, please do the testing process again!

sorry I know it's noisy but necessary.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,640 @@
         else
-          echo "Stopping mailcow..."
-          ${COMPOSE_COMMAND} -f ${COMPOSE_FILE} --env-file ${ENV_FILE} down
+          read -p "mailcow will be stopped and the currently active mailcow.conf will be modified to use the DB parameters found in ${RESTORE_LOCATION}/mailcow.conf - do you want to proceed? [Y|n] " MYSQL_STOP_MAILCOW
```
Here is actually a wrong path given to ouput: /opt/backup/mailcow-2024-08-27-08-14-51//mailcow.conf you see the double / ?

h3ssan replied:
This is a safe-guard, double slashes won't effect.

![](https://i.imgur.com/Cn2NwMw.png)

Put as many `/` as you need, will still works in bash !

DerLinkman reviewed (changes requested):
See comment

DerLinkman commented:
How does this script behave when it is automated? E.g runned via cron?

Especially the part with -d (delete backups) you have to enter YES to confirm deletion?

h3ssan commented:
> How does this script behave when it is automated? E.g runned via cron?
> 
> Especially the part with -d (delete backups) you have to enter YES to confirm deletion?

For now, it won't. In the next feat PR, I'll make it automated.

DerLinkman commented:
> For now, it won't. In the next feat PR, I'll make it automated.

That makes it hard for us to track... as we always take all staging commits into the next update and we await the full functionality of that. If you say that the automation is not implemented yet inside this pr, then we won't merge it... i'm sorry. As this is heavily used by other users.

In general: Why did you want to make multiple prs instead of one which includes all of your changes, even the new additions? I did not get it.

h3ssan commented:
> > For now, it won't. In the next feat PR, I'll make it automated.
> 
> That makes it hard for us to track... as we always take all staging commits into the next update and we await the full functionality of that. If you say that the automation is not implemented yet inside this pr, then we won't merge it... i'm sorry. As this is heavily used by other users.
> 
> In general: Why did you want to make multiple prs instead of one which includes all of your changes, even the new additions? I did not get it.

Then, If something happened unexpected, we will revert everything instead of the PR that has the problem. Please correct me if I am wrong.

As well, many features inside one PR will make the testing harder.

If you resist, then it's ok for me. Please confirm that you need the full implementation inside this PR and I'll start make the script fully automated.

DerLinkman commented:
I definitely unterstand your doing. I prefer this. But currently your backup script is removing a automated part. So yes: Please readd the automation into your scripts remake PR.

h3ssan committed:
Implement `--yes` flag, to automate `--delete` and `--backup` script

h3ssan committed:
Fix: `find: No such file or directory` when directory not contains data

h3ssan commented:
The script now can be automated using `--yes` flag.

Example:
```bash
./helper-scripts/backup_and_restore.sh --delete /opt/backup 7 --yes
```

> Please note that since `--yes` flag is critical and won't ask for any confirmation, therefore there will not be any short flag for it. Only `--yes`.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,670 @@
+        # Skip prompt if --yes flag is present
+        if [[ "${MAILCOW_YES_TO_ALL}" == "1" ]]; then
+          echo -e "${GREEN_COLOR}Forcing a resync now due to --yes flag.${RESET}"
+          docker exec $(docker ps -qf name=dovecot-mailcow) doveadm force-resync -A '*'
```
I tested the restoration on my machine with the --yes flag. However, it seems that after the mails are restored doveadm has problems to resync them if the command to resync them is directly fired afterwards. If a user enters this manually (y to restore) it won't throw an error, i think dovecot is searching for the userdb file to early:

```
Forcing a resync now due to --yes flag.
Error: auth-master: userdb list: connect(/run/dovecot/auth-userdb) failed: Connection refused
Panic: file auth-master.c: line 436 (auth_master_unset_io): assertion failed: (conn->to == NULL)
```

h3ssan replied:
Then, should `sleep 1` ? or there's any other idea like making sure a file is exists and then resync ?

DerLinkman replied:
Sleep 1 will mostlikely do it i guess, lets try that.

h3ssan replied:
Done, I've added `sleep 1s`, could you re-try again several times?

h3ssan replied:
Looks like this didn't fix it, make it 2seconds, now fixed.

h3ssan replied:
I tried several times, 2seconds solved it.

DerLinkman reviewed (changes requested):
See comment (only one :D) rest looks fine to me.

However, we need to adjust the mailcow docs too. Do you want to do it by yourself or should we adapt it?

Just asking, that we don't do the same work parallel :D

h3ssan commented:
> See comment (only one :D) rest looks fine to me.
> 
> However, we need to adjust the mailcow docs too. Do you want to do it by yourself or should we adapt it?
> 
> Just asking, that we don't do the same work parallel :D

Sure, I’ll do adjust the docs!

h3ssan committed:
Fix dovecot resync: failed: Connection refused

Full error message:
```
Error: auth-master: userdb list: connect(/run/dovecot/auth-userdb) failed: Connection refused
Panic: file auth-master.c: line 436 (auth_master_unset_io): assertion failed: (conn->to == NULL)
```

h3ssan committed:
Fix: dovecot resync: failed connection refused

DerLinkman commented:
@MAGICCC could you try to test this too? It works and looks good for me but if a seperate user can take a look at this i would be happy

MAGICCC commented:
Yes I will check on the weekend!

h3ssan commented:
> Yes I will check on the weekend!

Thank you!

mrclschstr commented:
Would be awesome to include this as well: https://github.com/mailcow/mailcow-dockerized/pull/4833

ashish1099 commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -1,331 +1,672 @@
+
+# Check for required files
 SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
 COMPOSE_FILE=${SCRIPT_DIR}/../docker-compose.yml
```
can we also accept compose file and env from env if possible, it would be nice to keep all of my compose in a set directory structure.

ashish1099 replied:
and ofcourse default should be the one, which we already have.

h3ssan replied:
> can we also accept compose file and env from env if possible, it would be nice to keep all of my compose in a set directory structure.

Hey @DerLinkman , please check is this possible/should be implemented?

DerLinkman replied:
No i don't like that. Sorry we have to keep mailcow as it is...

DerLinkman commented:
> Would be awesome to include this as well: #4833

@h3ssan could you adapt this in your next step (adding new features)?

h3ssan commented:
> > Would be awesome to include this as well: #4833
> 
> @h3ssan could you adapt this in your next step (adding new features)?

I went through the PR, interesting. I will implement it later (after completing this step).

milkmaker commented:
This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

mrclschstr commented:
Can this be reopened?

h3ssan commented:
> Can this be reopened?

This pull request cannot be reopened, as we have decided to keep `backup_and_restore.sh` as a legacy component. However, the good news is that we will develop a new solution with better options and capability.
